hnix-store-core/src/System/Nix/Nar.hs

```diff
@@ -63,217 +86,221 @@ data NarEffects (m :: * -> *) = NarEffects {
-        -- putContents fSize bs = str "contents" <> int (BSL.length bs) <> (pad fSize bs)
-
-        int :: Integral a => a -> B.Put
-        int n = B.putInt64le $ fromIntegral n
-
-        pad :: Int64 -> BSL.ByteString -> B.Put
-        pad strSize bs = do
-          B.putLazyByteString bs
-          B.putLazyByteString (BSL.replicate (padLen strSize) 0)
-
-        strs :: [BSL.ByteString] -> B.Put
-        strs = mapM_ str
-
-
-------------------------------------------------------------------------------
--- | Deserialize a Nar from lazy ByteString
-getNar :: B.Get Nar
-getNar = fmap Nar $ header >> parens getFile
-    where
-
-      header   = assertStr "nix-archive-1"
-
-
-      -- Fetch a FileSystemObject
-      getFile = getRegularFile <|> getDirectory <|> getSymLink
-
-      getRegularFile = do
-          assertStr "type"
-          assertStr "regular"
-          mExecutable <- optional $ Executable <$ (assertStr "executable"
-                                                   >> assertStr "")
-          assertStr "contents"
-          (fSize, contents) <- sizedStr
-          return $ Regular (fromMaybe NonExecutable mExecutable) fSize contents
-
-      getDirectory = do
-          assertStr "type"
-          assertStr "directory"
-          fs <- many getEntry
-          return $ Directory (Map.fromList fs)
-
-      getSymLink = do
-          assertStr "type"
-          assertStr "symlink"
-          assertStr "target"
-          fmap (SymLink . E.decodeUtf8 . BSL.toStrict) str
-
-      getEntry = do
-          assertStr "entry"
-          parens $ do
-              assertStr "name"
-              name <- E.decodeUtf8 . BSL.toStrict <$> str
-              assertStr "node"
-              file <- parens getFile
-              maybe (fail $ "Bad FilePathPart: " ++ show name)
-                    (return . (,file))
-                    (filePathPart $ E.encodeUtf8 name)
-
-      -- Fetch a length-prefixed, null-padded string
-      str = fmap snd sizedStr
-
-      sizedStr = do
-          n <- B.getInt64le
-          s <- B.getLazyByteString n
-          p <- B.getByteString . fromIntegral $ padLen n
-          return (n,s)
-
-      parens m = assertStr "(" *> m <* assertStr ")"
-
-      assertStr s = do
-          s' <- str
-          if s == s'
-              then return s
-              else fail "No"
-
-
--- | Distance to the next multiple of 8
-padLen :: Int64 -> Int64
-padLen n = (8 - n) `mod` 8
-
-
--- | Unpack a NAR into a non-nix-store directory (e.g. for testing)
-localUnpackNar :: Monad m => NarEffects m -> FilePath -> Nar -> m ()
-localUnpackNar effs basePath (Nar fso) = localUnpackFSO basePath fso
-
+isExecutable :: Functor m => NarEffects m -> FilePath -> m IsExecutable
+isExecutable effs fp =
+  bool NonExecutable Executable . Directory.executable <$> narGetPerms effs fp
+
+type NarParser m a = Parsec.ParsecT T.Text BSL.ByteString m a
+
+-- | Unpack a Nar bytestring into its filesystem objects
```

Comment:
Into FSOs? Or directly to disk?

Reply:
`unpackNarIO` writes straight to disk. Is that different from an FSO? Or were you asking if we unpack into the `FSO` type?